### DISCUSSION TITLE
The Road to Scala 3 support

### DISCUSSION THREAD

**Discussion by bishabosha** [slick/slick#2843] [Questions] [answered]
Hey, I saw on [Re-activate the scala-library-next initiative](https://contributors.scala-lang.org/t/re-activate-the-scala-library-next-initiative/6339) that @nafg says some libraries are blocking progress on releasing a Scala 3 version for Slick, so I wanted to ask what those are, or which issue/PR tracks these things, as I see https://github.com/slick/slick/pull/2687 was already merged? thanks

**Comment by nafg** [comment, marked as answer]
Hi, the libraries are slick-pg and slick-additions.

@mdedetrich starting working porting slick-pg: https://github.com/tminglei/slick-pg/pull/668 but I think he's away now.

For slick-additions, see https://github.com/nafg/slick-additions/pull/272. There are 3 issues:

1. The code generator is based on scalameta. I don't have a solution.
2. There is some reflection code which I attempted to replace with java reflection in https://github.com/nafg/slick-additions/pull/272/commits/a29a33f6f859a06ae4ae21d22fd93962b2c79e5a, not sure if it's correct but if it is then that's okay
3. There is some type inference issue, which I tried fixing, in the same commit

**Comment by nafg** [comment]
To be clear, there are already milestone releases of Slick with Scala 3 support. I just want to make sure before the final release that getting slick-pg and slick-additions to compile for Scala 3 won't require any breaking changes in Slick, that it will be too late to do after the final release.

**Comment by fabienfoerster** [comment]
Hi, it seems that both slick-pg and slick-additions compile for Scala 3. Is there something we are still waiting for before cutting the final release ? @nafg

**Comment by nafg** [comment]
I literally just got a working slick additions (pre) release. I want to also test it in real life. I would like to push it to production in the next day or two. Unless any risky pull requests have been merged since the last milestone I could then do a release, or an RC if people think there's any point.

There was also the flaky pekko test, not sure if that was solved @mdedetrich ?

Also I'd love to hear from people that use the latest milestone in production.

I'm really sorry it got pushed off so long. I've been insanely busy with a badly managed project.

So, I think potentially next week. Unless you've all been using it in prod already, then I don't necessarily need to wait to get to do so myself.

**Reply by mdedetrich** [reply]
> There was also the flaky pekko test, not sure if that was solved @mdedetrich ?

Ill have a look into this

**Reply by mdedetrich** [reply]
It solved itself, see https://github.com/slick/slick/issues/2811

**Comment by Lukah0173** [comment]
>Also I'd love to hear from people that use the latest milestone in production.

We're using `3.5.0-M5` in production with Scala 3.3.1. No issues and works great. There's some use of more complex patterns but mostly standard approaches. We're also migrating to Pekko Persistence JDBC, but in the meantime we've got this locked to Akka and Scala 2.13.x. Just let me know if I can provide any info with our use cases 👍

**Comment by nafg** [comment]
None of the PRs merged since M5 seem to be significant. Any opinions on RC vs. going straight to final?

**Reply by mdedetrich** [reply]
@nafg I have a quality of life improvement but its only applicable if release artifacts are built in CI (as opposed to on a dedicated local machine). Which is the case?

**Reply by nafg** [reply]
CI. What do you have in mind?

**Reply by mdedetrich** [reply]
I have a PR that enables the Scala 2 inliner only in CI as it can cause issues with local development due to zinc/incremental compilation. Will submit the PR now.

**Reply by mdedetrich** [reply]
I also noticed https://github.com/slick/slick/pull/2301 , is it worth it to finish it for 3.5.0 or should it just be delayed for 3.6.0?

**Reply by nafg** [reply]
I think it should be a later version (orthogonal to how soon it's finished and released; 3.6.0 could be released as soon as there's reason to). I think it's good to have minimal changes other than Scala 3 support, although of course whatever changes were done already are good to have.

I mean if someone would implement it tomorrow it could go into 3.5.0 but we shouldn't wait for it to release it.

**Reply by nafg** [reply]
(For the same reason, once 3.5.0 is released I would feel comfortable experimenting and enhancing Slick generally, knowing that if people can get Scala 3 support without any unnecessary migration work.)

**Reply by PerWiklander** [reply]
Today was my monthly "Is Slick on Scala 3 yet?" check in. I've been doing that since around 2021.
It looks like I can start getting my hope up now :-)

Slick is the last blocker for our Scala 3 migration (I am assuming other eco system libraries like akka-`persistence-jdbc` and `akka-projection-slick` will follow shortly).

**Reply by nafg** [reply]
I just tagged RC1 (but if publishing fails requiring pushing a fix, I may delete & retag)

**Comment by nafg** [comment]
RC1 is on Maven Central!

Please test and report your experiences (positive or negative), here or in https://github.com/slick/slick/discussions/2891.

Also, contributions to https://github.com/slick/slick/blob/v3.5.0-RC1/doc/paradox/upgrade.md would be appreciated.

**Reply by mkurz** [reply]
New play-slick releases published:
- [5.3.0-RC1](https://github.com/playframework/play-slick/releases/tag/5.3.0-RC1)
- [6.1.0-RC1](https://github.com/playframework/play-slick/releases/tag/6.1.0-RC1)
